source/common/protobuf/utility.cc

```diff
@@ -109,6 +111,50 @@ void jsonConvertInternal(const Protobuf::Message& source,
   MessageUtil::loadFromJson(json, dest, validation_visitor);
 }
 
+// Apply a function transforming a message (e.g. loading JSON into the message).
+// First we try with the message's earlier type, and if unsuccessful (or no
+// earlier) type, then the current type. This allows us to take a v3 Envoy
+// internal proto and ingest both v2 and v3 in methods such as loadFromJson.
+// This relies on the property that any v3 configuration that is readable as
+// v2 has the same semantics in v2/v3, which holds due to the highly structured
+// vN/v(N+1) mechanical transforms.
+void tryWithApiBoosting(std::function<void(Protobuf::Message&, bool)> f,
```

Comment:
nit: prefer dual state enum to `bool` param to lambda. Very hard to read what is going on here.